storage/pebble/procedure/children.go

```diff
@@ -11,11 +12,30 @@ import (
 	"github.com/onflow/flow-go/storage/pebble/operation"
 )
 
+type BlockIndexer struct {
+	// indexing is a mutex to avoid dirty reads when calling RetrieveBlockChildren
+	indexing *sync.Mutex
+}
+
+func NewBlockIndexer() *BlockIndexer {
+	return &BlockIndexer{
+		indexing: new(sync.Mutex),
+	}
+}
+
+var _ storage.BlockIndexer = (*BlockIndexer)(nil)
+
+func (bi *BlockIndexer) IndexNewBlock(blockID flow.Identifier, parentID flow.Identifier) func(storage.PebbleReaderBatchWriter) error {
+	return IndexNewBlock(bi.indexing, blockID, parentID)
+}
+
 // IndexNewBlock will add parent-child index for the new block.
 //   - Each block has a parent, we use this parent-child relationship to build a reverse index
 //   - for looking up children blocks for a given block. This is useful for forks recovery
 //     where we want to find all the pending children blocks for the lastest finalized block.
 //
+// # It's concurrent safe to call this function by multiple goroutines, as it will acquire a lock
```

Comment:
Suppose we have 2 threads A and B concurrently writing a block that initially has no children, the sequence diagram would look like:
```mermaid
sequenceDiagram
    par
        A ->> DB: starts write batch
    and
        B ->> DB: starts write batch
    end
    A -> DB: acquire lock
    activate A
    A -> DB: read Children (nil)
    A -> DB: write Chidren [c1]
    A -> DB: release lock
    deactivate A
    B -> DB: acquire lock
    activate B
    B -> DB: read Children (nil)
    B -> DB: write Chidren [c2]
    B -> DB: release lock
    deactivate B
    par
        A ->> DB: commit write batch
    and
        B ->> DB: commit write batch
    end
  ```

The lock added here does not provide any guarantee that this function will operate correctly when called in parallel, **because the lock is released before the write is committed**. 

Suppose we have some database operation `P` consisting of a set of reads `R` and a set of writes `W`, where the writes are dependent on the values read in `R`. In general, to safely synchronize `P`, I think we need to:
- acquire `Lock_P`, the lock protecting access to `P`
- perform all reads in `R`, and memorize the values
- instantiate the write batch
- perform all writes in `W`
- commit the write batch
- release `Lock_P`

Reply:
Nice diagram. And thanks for the feedback.

[It's addressed here](https://github.com/onflow/flow-go/pull/6360/commits/77b64c0f38644eda6abb593025e410ec8c2f8a6e)

I used the callback to release the lock after the batch update is committed. 

Also added a test case to reproduce and verify the fix.